vortex-dtype/src/arrow.rs

```diff
@@ -189,34 +349,121 @@ impl DType {
-                l.to_arrow_dtype()?,
-                l.nullability().into(),
+            DType::List(elem_type, _) => DataType::List(FieldRef::new(Field::new_list_field(
+                elem_type.to_arrow()?,
+                elem_type.nullability().into(),
             ))),
             DType::Extension(ext_dtype) => {
                 // Try and match against the known extension DTypes.
                 if is_temporal_ext_type(ext_dtype.id()) {
                     make_arrow_temporal_dtype(ext_dtype)
                 } else {
-                    vortex_bail!("Unsupported extension type \"{}\"", ext_dtype.id())
+                    ext_dtype.storage_dtype().to_arrow()?
                 }
             }
         })
     }
 }
 
+/// If a suitable [conversion is registered][DType::register_extension_type], returns the Arrow `Field`
+/// representing that this extension type converts into. Otherwise, `None` is returned.
+///
+/// Any errors from the conversion function are propagated here via
+/// the outer `VortexResult` wrapper.
+pub fn extension_type_to_arrow(ext_type: &ExtDType) -> VortexResult<Option<Field>> {
+    for kernel in EXTENSION_TYPES.read().vortex_expect("poisoned").iter() {
+        if let Some(metadata) = kernel.0.to_arrow(ext_type)? {
+            return Ok(Some(metadata));
+        }
+    }
+
+    Ok(None)
+}
+
+/// Convert an Arrow [`Field`] to a `DType` using a [registered extension][ArrowTypeConversion].
+///
+/// If no suitable conversion has been registered, `Ok(None)` is returned.
+pub fn arrow_field_to_dtype(field: &Field) -> VortexResult<Option<DType>> {
+    for converter in EXTENSION_TYPES.read().vortex_expect("poisoned").iter() {
+        if let Some(converted) = converter.to_vortex(field)? {
+            return Ok(Some(converted));
+        }
+    }
+
+    Ok(None)
+}
+
+/// Conversions between Arrow [`Field`] type and Vortex logical `DType`.
+///
+/// This crate provides infra for registering and discovering extension types.
+/// Once you implement this trait, you need to register it using [`DType::register_extension_type`]
+/// to have it get discovered.
+///
+/// See also: [`DType::register_extension_type`]
+pub trait ArrowTypeConversion: 'static + Send + Sync + Debug {
+    /// Convert the given Arrow [`Field`] to a Vortex [`DType`].
+    fn to_vortex(&self, _field: &Field) -> VortexResult<Option<DType>> {
+        Ok(None)
+    }
+
+    /// Map a type into an Arrow [`Field`] type.
+    ///
+    /// This method should be implemented if you want to add support for a Vortex
+    /// [extension type][ExtDType] that also has a corresponding canonical Arrow
+    /// extension type, to preserve round-trip between the two.
+    #[allow(clippy::disallowed_types)]
+    fn to_arrow(&self, _dtype: &ExtDType) -> VortexResult<Option<Field>> {
+        Ok(None)
+    }
+}
+
+/// Conversion token
+#[derive(Debug)]
+pub struct ArrowTypeConversionRef(pub ArcRef<dyn ArrowTypeConversion>);
```

Comment:
We tend to just type alias `XXRef` things. Then you get all the other traits impl'd by ArcRef.

Reply:
I patterned this off of the XYZKernelRef structs which all do this, unless something has changed in the last 48 hours